cpp/src/arrow/type.h

```diff
@@ -1582,13 +1583,26 @@ class ARROW_EXPORT FieldRef {
 // ----------------------------------------------------------------------
 // Schema
 
+enum class Endianness {
+  LITTLE = 0,
+  BIG = 1,
+#if ARROW_LITTLE_ENDIAN
+  NATIVE = LITTLE
+#else
+  NATIVE = BIG
+#endif
+};
+
 /// \class Schema
 /// \brief Sequence of arrow::Field objects describing the columns of a record
 /// batch or table data structure
 class ARROW_EXPORT Schema : public detail::Fingerprintable,
                             public util::EqualityComparable<Schema>,
                             public util::ToStringOstreamable<Schema> {
  public:
+  explicit Schema(std::vector<std::shared_ptr<Field>> fields, Endianness endianness,
+                  std::shared_ptr<const KeyValueMetadata> metadata = NULLPTR);
```

Comment:
I don't know if we want to allow non-native endianness in memory. @wesm What do you think?

Reply:
cc @kou

Reply:
Here are use cases I imagine:

  * Router: It doesn't process any Arrow data and just passes the received Arrow data to another system. In this case,  it's efficient that we allow non-native endianness in memory.
  * Saver: It doesn't process any Arrow data and just saves the received Arrow data to filesystem. For example, it saves the received Arrow data to S3 as is.

Does it make sense?

Reply:
@kou Yes, that seems to make sense.